its/src/test/java/com/sonar/it/scanner/msbuild/sonarcloud/JreProvisioningTest.java

```diff
@@ -0,0 +1,115 @@
+import com.sonar.orchestrator.util.CommandExecutor;
+import com.sonar.orchestrator.util.StreamConsumer;
+import java.io.File;
+import java.io.IOException;
+import java.io.StringWriter;
+import java.nio.file.Path;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.io.TempDir;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import static org.assertj.core.api.Assertions.assertThat;
+
+class JreProvisioningTest {
+  private final static Logger LOG = LoggerFactory.getLogger(JreProvisioningTest.class);
+  private final static Integer COMMAND_TIMEOUT = 2 * 60 * 1000;
+  private final static String SCANNER_PATH = "../build/sonarscanner-net-framework/SonarScanner.MSBuild.exe";
+  private final static String SONARCLOUD_PROJECT_KEY = "team-lang-dotnet_jre-provisioning";
+  private final static String PROJECT_NAME = "JreProvisioning";
+
+  @TempDir
+  public Path basePath;
+
+  @Test
+  void different_hostUrl_sonarcloudUrl_logsAndExitsEarly() {
+    var logWriter = new StringWriter();
+    StreamConsumer.Pipe logsConsumer = new StreamConsumer.Pipe(logWriter);
+
+    var beginCommand = Command.create(new File(SCANNER_PATH).getAbsolutePath())
+      .addArgument("begin")
+      .addArgument("/o:org")
+      .addArgument("/k:project")
+      .addArgument("/d:sonar.host.url=http://localhost:4242")
+      .addArgument("/d:sonar.scanner.sonarcloudUrl=" + Constants.SONARCLOUD_URL);
+
+    LOG.info("Scanner path: " + SCANNER_PATH);
+    LOG.info("Command line: " + beginCommand.toCommandLine());
+    var beginResult = CommandExecutor.create().execute(beginCommand, logsConsumer, COMMAND_TIMEOUT);
+    assertThat(beginResult).isOne();
+
+    assertThat(logWriter.toString()).contains(
+      "The arguments 'sonar.host.url' and 'sonar.scanner.sonarcloudUrl' are both set and are different. Please set either 'sonar.host" +
+        ".url' for SonarQube or 'sonar.scanner.sonarcloudUrl' for SonarCloud.");
+  }
+
+  @Test
+  void jreProvisioning_skipProvisioning_doesNotDownloadJre() throws IOException {
+    var projectDir = TestUtils.projectDir(basePath, PROJECT_NAME);
+    var logWriter = new StringWriter();
+    StreamConsumer.Pipe logsConsumer = new StreamConsumer.Pipe(logWriter);
+
+    SonarCloudUtils.runBeginStep(projectDir,SONARCLOUD_PROJECT_KEY, logsConsumer, "/d:sonar.scanner.skipJreProvisioning=true");
+
+    assertThat(logWriter.toString()).contains(
+      "JreResolver: Resolving JRE path.",
+      "JreResolver: sonar.scanner.skipJreProvisioning is set, skipping JRE provisioning.");
+    assertThat(logWriter.toString()).doesNotContain(
+      "JreResolver: Cache miss.",
+      "JreResolver: Cache hit",
+      "JreResolver: Cache failure.");
+  }
+
+  @Test
+  void jreProvisioning_endToEnd_cacheMiss_downloadsJre() throws IOException {
+    var projectDir = TestUtils.projectDir(basePath, PROJECT_NAME);
+
+    var logs = SonarCloudUtils.runAnalysis(projectDir, SONARCLOUD_PROJECT_KEY, "/d:sonar.userHome=" + projectDir.toAbsolutePath());
```

Comment:
This is the meat and potatoes of the PR.
It is also the only test that sets `sonar.userHome`, as the other tests maybe benefit from caching so that the pipeline is not too slow.

I want to add one more test with a cache hit, but I'll do it next week.